library/Backends/llvm/src/Juvix/Backends/LLVM/Codegen/Types.hs

```diff
@@ -0,0 +1,308 @@
+{-# LANGUAGE ConstraintKinds #-}
+
+module Juvix.Backends.LLVM.Codegen.Types
+  ( module Juvix.Backends.LLVM.Codegen.Types,
+    module Juvix.Backends.LLVM.Codegen.Types.Shared,
+  )
+where
+
+import qualified Data.ByteString.Short as Short hiding (empty)
+import qualified Distribution.System as System
+import Juvix.Backends.LLVM.Codegen.Types.Shared
+import Juvix.Library hiding (Type)
+import qualified Juvix.Library.HashMap as Map
+import LLVM.AST as AST
+import qualified LLVM.AST.AddrSpace as AddrSpace
+import LLVM.AST.DataLayout (DataLayout (..))
+import Prelude ((!!))
+
+--------------------------------------------------------------------------------
+-- Codegen State
+--------------------------------------------------------------------------------
+
+data CodegenState = CodegenState
+  { -- | Name of the active block to append to
+    currentBlock :: Name,
+    -- | Blocks for function
+    blocks :: Map.T Name BlockState,
+    -- | Function scope symbol table
+    symTab :: SymbolTable,
+    -- | Count of basic blocks
+    blockCount :: Int,
+    -- | Count of unnamed instructions
+    count :: Word,
+    -- | Name Supply
+    names :: Names,
+    moduleAST :: AST.Module,
+    -- | Debug level
+    debug :: Int
+  }
+  deriving (Show, Generic)
+
+data BlockState = BlockState
+  { -- | Block index
+    idx :: Int,
+    -- | Stack of instructions
+    stack :: [Named Instruction],
+    -- | Block terminator
+    term :: Maybe (Named Terminator)
+  }
+  deriving (Show, Generic)
+
+data Errors
+  = -- | Error when a block does not exist
+    NoSuchBlock Text
+  | -- | Error when a Variant does not exist
+    NoSuchVariant Text
+  | -- | Error that should never happen
+    DoesNotHappen Text
+  | -- | Error that happens when a variable out of scope is called
+    VariableNotInScope Text
+  | -- | Error that happens when a block lacks a terminator when it should have one
+    BlockLackingTerminator Int
```

Comment:
Where do some of these errors come from?

Reply:
The first one happens when the variable is not in the map when we wish to call it, so we throw variable not in scope erorr.

In the second case, LLVM demands every block has a terminator of some kind (either in a ret or a jump), and thus this occurs if we don't generate one